File: ide/src/main/java/org/dau/ide/project/ProjectTabs.java
package org.dau.ide.project;

import javafx.collections.SetChangeListener;
import javafx.scene.control.TabPane;
import org.dau.ide.schema.SchemaTab;
import org.dau.ui.schematic.model.FxProject;
import org.dau.ui.schematic.model.FxSchema;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

@Component
public class ProjectTabs extends TabPane {

  public ProjectTabs(ProjectManagementTab settingsTab) {
    super(settingsTab);
  }

  @Autowired
  public void initSchemas(FxProject project) {
    project.getSchemas().addListener((SetChangeListener.Change<? extends FxSchema> c) -> {
      if (c.wasRemoved()) {
        var schema = c.getElementRemoved();
        getTabs().removeIf(tab -> tab instanceof SchemaTab t && t.getSchema() == schema);
      }
    });
  }
}

File: ide/src/main/kotlin/org/dau/ide/project/ProjectTabs.kt
package org.dau.ide.project

import javafx.collections.SetChangeListener
import javafx.scene.control.TabPane
import org.dau.ide.schema.SchemaTab
import org.dau.ui.schematic.model.FxProject
import org.dau.ui.schematic.model.FxSchema
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

@Component
class ProjectTabs(settingsTab: ProjectManagementTab) : TabPane(settingsTab) {

  @Autowired
  fun initSchemas(project: FxProject) {
    project.schemas.addListener { c: SetChangeListener.Change<out FxSchema> ->
      if (c.wasRemoved()) {
        val schema = c.elementRemoved
        tabs.removeIf { it is SchemaTab && it.schema === schema }
      }
    }
  }
}
